src/Umbraco.Infrastructure/DependencyInjection/UmbracoBuilder.Uniques.cs

```diff
@@ -109,18 +109,53 @@ public static void SetShortStringHelper(this IUmbracoBuilder builder, IShortStri
         }
 
         /// <summary>
-        /// Sets the underlying media filesystem.
+        /// Sets the filesystem used by the MediaFileManager
         /// </summary>
         /// <param name="builder">A builder.</param>
-        /// <param name="filesystemFactory">A filesystem factory.</param>
+        /// <param name="filesystemFactory">Factory method to create an IFileSystem implementation used in the MediaFileManager</param>
+        public static void SetMediaFileSystem(this IUmbracoBuilder builder,
+            Func<IServiceProvider, IFileSystem> filesystemFactory) => builder.Services.AddSingleton(
+            provider =>
+            {
+                IFileSystem filesystem = filesystemFactory(provider);
+                // We need to use the Filesystems to create a shadow wrapper,
+                // because shadow wrapper requires the IsScoped delegate from the FileSystems.
+                // This is used by the scope provider when taking control of the filesystems.
+                FileSystems fileSystems = provider.GetRequiredService<FileSystems>();
+                IFileSystem shadow = fileSystems.CreateShadowWrapper(filesystem, "media");
+
+                return provider.CreateInstance<MediaFileManager>(shadow);
+            });
+
+        /// <summary>
+        /// Register FileSystems with a specific IFileSystem for stylesheets
+        /// </summary>
         /// <remarks>
-        /// Using this helper will ensure that your IFileSystem implementation is wrapped by the ShadowWrapper
+        /// Be careful when using this, the root path and root url must be correct for this to work.
         /// </remarks>
-        public static void SetMediaFileSystem(this IUmbracoBuilder builder, Func<IServiceProvider, IFileSystem> filesystemFactory)
-            => builder.Services.AddUnique<IMediaFileSystem>(factory =>
+        /// <param name="builder">A builder.</param>
+        /// <param name="stylesheetFactory">Factory method to create an IFileSystem implementation used for stylesheets.</param>
+        /// <exception cref="InvalidOperationException">Throws exception if full path can't be resolved successfully.</exception>
+        public static void AddFileSystems(this IUmbracoBuilder builder,
```

Comment:
Any reason this is called AddFileSystems when it's only for stylesheets? IMO the naming should be consistent like the media one: `SetStylesheetFileSystem` ?

Reply:
I chose to call it AddFileSystems because that's what the extension method does, but I can see that it's a bit confusing that we use AddFileSystems to change the stylesheet filesystem.